lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,292 @@
+  }
+
+  /*
+   * Given a rule, check if it's validate
+   */
+  bool Houdini::validateRule(HornRule r, HornClauseDB &db, HornifyModule &hm)
+  {
+	  Expr ruleHead_cand_app = applyActualArgsToCand(r.head());
+	  Expr neg_ruleHead_cand_app = mk<NEG>(ruleHead_cand_app);
+
+	  Expr ruleBody = r.body();
+	  ExprMap body_map;
+	  ExprVector body_pred_apps;
+	  get_all_pred_apps(ruleBody, db, std::back_inserter(body_pred_apps));
+
+	  for(ExprVector::iterator itr = body_pred_apps.begin(); itr != body_pred_apps.end(); ++itr)
+	  {
+		  body_map.insert(std::pair<Expr, Expr>(*itr, applyActualArgsToCand(*itr)));
+	  }
+
+	  Expr body_cand_app = replace(ruleBody, body_map);
+
+	  Expr whole_cand = mk<AND>(neg_ruleHead_cand_app, body_cand_app);
+
+	  errs() << "WHOLE CANDIDATE: " << *whole_cand << "\n";
+
+	  ZSolver<EZ3> solver(hm.getZContext ());
+	  solver.assertExpr(whole_cand);
+	  solver.toSmtLib(errs());
+	  bool isSat = solver.solve();
+	  if(isSat)
+	  {
+		  errs() << "SAT\n";
+	  }
+	  else
+	  {
+	  	  errs() << "UNSAT\n";
+	  }
+	  return isSat;
+  }
+
+  /*
+   * Given a function application, apply its actual arguments to its candidate
+   */
+  Expr Houdini::applyActualArgsToCand(Expr fapp)
+  {
+	  Expr fdecl = bind::fname(fapp);
+	  Expr cand = currentCandidates[fdecl];
+
+	  ExprMap bvar_map;
+	  ExprVector bvars;
+	  get_all_bvars(cand, std::back_inserter(bvars));
+
+	  for(ExprVector::iterator it = bvars.begin(); it != bvars.end(); ++it)
+	  {
+		  unsigned bvar_id = bind::bvarId(*it);
+		  Expr app_arg = bind::domainTy(fapp, bvar_id);// To improve
+		  bvar_map.insert(std::pair<Expr,Expr>(*it, app_arg));
+	  }
+	  Expr cand_app = replace(cand, bvar_map);
+	  return cand_app;
+  }
+
+  /*
+   * Given a rule, weaken its head's candidate
+   */
+  void Houdini::weakenRuleHeadCand(HornRule r)
+  {
+	  Expr ruleHead_app = r.head();
+	  Expr ruleHead_rel = bind::fname(ruleHead_app);
+	  Expr ruleHead_cand = currentCandidates[ruleHead_rel];
+
+	  if(isOpX<TRUE>(ruleHead_cand))
+	  {
+	  		return;
+	  }
+	  if(!isOpX<AND>(ruleHead_cand))
+	  {
+	  		Expr weaken_cand = mk<TRUE>(ruleHead_cand->efac());
+	  		currentCandidates[ruleHead_rel] = weaken_cand;
+	  }
+	  else
+	  {
+	  		ExprVector head_cand_args;
+	  		head_cand_args.insert(head_cand_args.end(), ruleHead_cand->args_begin(), ruleHead_cand->args_end());
+	  		head_cand_args.erase(head_cand_args.begin());
+
+	  		if(head_cand_args.size() > 1)
+	  		{
+	  			Expr weaken_cand = mknary<AND>(head_cand_args.begin(), head_cand_args.end());
+	  			currentCandidates[ruleHead_rel] = weaken_cand;
+	  		}
+	  		else
+	  		{
+	  			Expr weaken_cand = head_cand_args[0];
+	  			currentCandidates[ruleHead_rel] = weaken_cand;
+	  		}
+	  }
+  }
+
+  /*
+   * Given a rule head, extract all rules using it in body, then add all such rules to the end of workList
+   */
+  HornClauseDB::RuleVector Houdini::addUsedRulesBackToWorkList(HornClauseDB &db, HornClauseDB::RuleVector workList, Expr ruleHead_app)
+  {
```

Comment:
You can use set for uniqueness and list for order and update both as
needed.
On Tue, Jun 28, 2016 at 15:14 Chenguang-Zhu <notifications@github.com>
wrote:

> In lib/seahorn/Houdini.cc
> <https://github.com/seahorn/seahorn/pull/33#discussion_r68823360>:
>
> > +	  			Expr weaken_cand = mknary<AND>(head_cand_args.begin(), head_cand_args.end());
> > +	  			currentCandidates[ruleHead_rel] = weaken_cand;
> > +	  		}
> > +	  		else
> > +	  		{
> > +	  			Expr weaken_cand = head_cand_args[0];
> > +	  			currentCandidates[ruleHead_rel] = weaken_cand;
> > +	  		}
> > +	  }
> > +  }
> > +
> > +  /*
> > +   * Given a rule head, extract all rules using it in body, then add all such rules to the end of workList
> > +   */
> > +  HornClauseDB::RuleVector Houdini::addUsedRulesBackToWorkList(HornClauseDB &db, HornClauseDB::RuleVector workList, Expr ruleHead_app)
> > +  {
>
> std::set will sort its content automatically. I'm afraid by using that
> when we push some rules back to worklist it will disorder the rules.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/seahorn/seahorn/pull/33/files/cce7a302670f7c9e1a2207c15b3c5d1ac93ec881#r68823360>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFzN50-VeJ6gp1iDsrndcTPeKuBTYVZ2ks5qQXKngaJpZM4I7V-j>
> .
>